### PR TITLE
Update asset-swapper to support MultiplexFeature

### PR DESCRIPTION
**Pull request by mzhu25**
## Description
This PR makes the necessary changes for AssetSwapper to utilize `MultiplexFeature`, and to account for the gas savings afforded by `MultiplexFeature` in routing (note that  we'll also need to update `exchangeProxyOverhead` in the 0x API config). 
The `ExchangeProxySwapQuoteConsumer` checks whether a swap quote can utilize `Multiplex` and converts the quote into `Multiplex` calldata if so. 
There's also some hackery around the source flags so that the `exchangeProxyOverhead` function can distinguish between RFQ and limit orders, since `MultiplexFeature.batchFill` only supports RFQ. 

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

**Comment by mzhu25** [comment]
Running some preliminary tests on simbot, this is a split between Uniswap [WETH, DAI] and Uniswap [WETH, USDT, DAI]:
```
SELL WETH->DAI 44.91 -> 79129.13 ($79833.80) after 86.8s @ http://localhost:7001/swap/dev/quote
	✔ PASS 44.91 -> 79056.86 ($78813.51) ($41.84)
	Uniswap_V2: 100%
	gas: 219438
```

**Comment by mzhu25** [comment]
here's a Uni/Sushi split:
```
SELL DAI->WETH 247905.17 -> 139.14 ($247273.26) after 67.4s @ http://localhost:7001/swap/dev/quote
	✔ PASS 247905.17 -> 139.13 ($247323.66) ($36.11)
	Uniswap_V2: 46%, SushiSwap: 54%
	gas: 185508
```

**Review by dekz — APPROVED** [review]

File: packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 1]
```diff
@@ -36,28 +37,49 @@ import {
...
+    { name: 'signature', type: 'tuple', components: SIGNATURE_ABI },
+]);
+const uniswapDataEncoder = AbiEncoder.create([
+    { name: 'tokens', type: 'address[]' },
+    { name: 'isSushi', type: 'bool' },
+]);
+const plpDataEncoder = AbiEncoder.create([
+    { name: 'provider', type: 'address' },
+    { name: 'auxiliaryData', type: 'bytes' },
+]);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts:81]: Might be clearer to namespace or prefix these for multiplex.

File: packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 2]
```diff
@@ -380,91 +421,147 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
...
-    }
-    const order = quote.orders[0];
-    if (!directSources.includes(order.source)) {
-        return false;
-    }
-    // VIP does not support selling the entire balance
-    if (opts.shouldSellEntireBalance) {
-        return false;
+    private _encodeMultiplexBatchFillCalldata(quote: SwapQuote): string {
+        const multiplex = new MultiplexFeatureContract(NULL_ADDRESS, this.provider);
```
Comment (dorothy-zbornak) [inline packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts:426]: Not sure how heavy these contract objects are but might ease the load to create them once in the constructor.

**Review by dorothy-zbornak — APPROVED** [review]

File: packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts [hunk 3]
```diff
@@ -380,91 +421,147 @@ export class ExchangeProxySwapQuoteConsumer implements SwapQuoteConsumerBase {
...
+                    ];
+                    wrappedBatchCalls.push({
+                        selector: this._exchangeProxy.getSelector('_transformERC20'),
+                        sellAmount: BigNumber.sum(...quote.orders.slice(i).map(o => o.takerAmount)),
+                        data: transformERC20Encoder.encode({
+                            transformations,
+                            ethValue: constants.ZERO_AMOUNT,
+                        }),
+                    });
+                    break for_loop;
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply packages/asset-swapper/src/quote_consumers/exchange_proxy_swap_quote_consumer.ts:493]: yeah we don't enforce it in the contract

File: packages/asset-swapper/src/utils/market_operation_utils/constants.ts [hunk 4]
```diff
@@ -102,9 +102,16 @@ export const PROTOCOL_FEE_MULTIPLIER = new BigNumber(70000);
...
+// HACK(mzhu25): Limit and RFQ orders need to be treated as different sources
+//               when computing the exchange proxy gas overhead.
+export const SOURCE_FLAGS: { [source in ERC20BridgeSource]: number } & {
+    RfqOrder: number;
+    LimitOrder: number;
+} = Object.assign(
     {},
-    ...Object.values(ERC20BridgeSource).map((source: ERC20BridgeSource, index) => ({ [source]: 1 << index })),
+    ...['RfqOrder', 'LimitOrder', ...Object.values(ERC20BridgeSource)].map(
+        (source: ERC20BridgeSource | 'RfqOrder' | 'LimitOrder', index) => ({ [source]: 1 << index }),
```
[the remark this answers is not in the document]
Comment (mzhu25) [reply packages/asset-swapper/src/utils/market_operation_utils/constants.ts:113]: sure I'll set it to 0

**Pull request merged**